### PR TITLE
Fixed #29408 -- Added validation of related fields and lookups in model Meta.ordering.

### PR DESCRIPTION
**Pull request by hramezani**
[ticket 29408](https://code.djangoproject.com/ticket/29408)

### CODE REVIEW

File: tests/foreign_object/models/article.py [hunk 1]
```diff
@@ -87,7 +87,7 @@ class ArticleTranslation(models.Model):
 
     class Meta:
         unique_together = ('article', 'lang')
-        ordering = ('active_translation__title',)
+        # ordering = ('active_translation__title',)
```
Comment (hramezani) [inline tests/foreign_object/models/article.py]: @timgraham I had to comment this line. because it broke the tests. any idea?

Comment (timgraham) [reply tests/foreign_object/models/article.py]: Seems fine to remove. `ArticleTranslation.objects.all()` fails with `FieldError: Cannot resolve keyword 'active_translation' into field.`. It looks like that was always broken.

Comment (hramezani) [reply tests/foreign_object/models/article.py]: @timgraham I removed the line

**Comment by hramezani** [comment]
@timgraham I will try it as soon as possible.

**Comment by timgraham** [comment]
Also, the new code seems somewhat complicated. Retrieving a field from a path seems like a common use case which might warrant a new method in `django.db.models.options`.  Did you adapt this code from another place in Django? If not, searching for LOOKUP_SEP might be the way to find similar code to help determine what a suitable API might be, something like `get_field_by_path()`, perhaps). That could help simplify this to something like:
```python
for field in fields:
try:
    cls._meta.get_field_by_path(field)
except FieldDoesNotExist:
   errors.append(...)
```

**Comment by hramezani** [comment]
@timgraham thank you for the comment.
I am a little confused about your comment. I added some code to check ordering field existence in related models by any depths. do you mean to add a method for retrieve field by path in any depth and pass the related field to this method for checking?

sorry for my delay in response, I am in process of moving to German.

**Comment by timgraham** [comment]
The related issue is about JSON property lookups like `'json_field__property` where property is a value in the JSON data.

**Comment by hramezani** [comment]
@timgraham I added `get_field_by_path` method to `django.db.models.options`. sorry for the dalay.

**Comment by hramezani** [comment]
@timgraham I added the `get_field_by_path()` to `django.db.models.options`. can you check it?
please let me know if anything else remains. Thanks

File: tests/foreign_object/models/article.py [hunk 2]
```diff
@@ -87,7 +87,6 @@ class ArticleTranslation(models.Model):
 
     class Meta:
         unique_together = ('article', 'lang')
-        ordering = ('active_translation__title',)
```
Comment (carltongibson) [inline tests/foreign_object/models/article.py]: Why this change?

File: django/db/models/options.py [hunk 3]
```diff
@@ -566,6 +571,26 @@ def get_field(self, field_name):
         except KeyError:
             raise FieldDoesNotExist("%s has no field named '%s'" % (self.object_name, field_name))
 
+    def get_field_by_path(self, path):
```
Comment (carltongibson) [inline django/db/models/options.py]: If we're going to add this, is there an opportunity to share the logic, rather than just copy from `get_fields_by_path()`? ("I've seen this before...", I thought.)

Comment (hramezani) [reply django/db/models/options.py]: it is from [get_fields_from_path](https://github.com/django/django/blob/da92ec79621fc0bba671d8afa52b7f6884962fe5/django/contrib/admin/utils.py#L482) in `contrib/admin/utils.py`

File: tests/foreign_object/models/article.py (hunk 2 above)
Comment (hramezani) [reply tests/foreign_object/models/article.py]: @timgraham comment:

> Seems fine to remove. ArticleTranslation.objects.all() fails with FieldError: Cannot resolve keyword 'active_translation' into field.. It looks like that was always broken.

**Comment by hramezani** [comment]
@carltongibson thank you for the comments. I am going to investigate.

**Comment by charettes** [comment]
As @carltongibson mentioned I this seems to disallow transforms usages in `Meta.ordering` and not only for JSON fields (e.g. that would disallow `field__lower`). The only way to perform a correct validation of values allowed to be passed to `Meta.ordering` is to rely on `get_lookups` and `get_transforms` similarly to what `Query.build_lookup()` does https://github.com/django/django/blob/f3d3338e06d571a529bb2046428eeac8e56bcbf6/django/db/models/sql/query.py#L1089-L1132. I don't this warrant the introduction of a new `Meta.get_field_by_path` method.

**Comment by charettes** [comment]
To summarize the logic should be as follow.

1. Split the lookup by `LOOKUP_SEP`
2. Do a `get_field` on the first part of the splitted lookup, that will be your `lhs`.
3. For the remaining parts do `transform_cls = lhs.get_transform(part); lhs = transform_cls(lhs)`.

**Comment by hramezani** [comment]
@carltongibson , @charettes thanks for the comments.

First of all, I rebased my PR with the master.

@carltongibson the changes also cover your comment about the `JSON Paths`. Also I added a test for it.

@charettes for the case that you proposed in your comment (`field__lower`). Now we can't order a query by this.
for example this query will cause error:
`Model.objects.all().order_by('char_field__lower')`

**Comment by collinanderson** [comment]
Does this handle multiple levels deep of json? `jsonfield__object__value`? (not sure anyone would actually want to do that, or if ordering by that is even possible.)

File: django/db/models/base.py [hunk 4]
```diff
@@ -1630,9 +1630,39 @@ def _check_ordering(cls):
...
+                    fld = _cls._meta.get_field(part)
+                    if fld.is_relation:
+                        _cls = fld.get_path_info()[-1].to_opts.model
+                    else:
+                        _cls = None
+                except (FieldDoesNotExist, AttributeError):
+                    if fld is None or fld.get_transform(part) is None:
+                        errors.append(
+                            checks.Error(
+                                "Invalid 'ordering' item '%s'." % field,
```
Comment (carltongibson) [inline django/db/models/base.py]: I think this should probably be `"'ordering' refers to the nonexistent related field or lookup '%s'." % field` to match the other error messages.

File: tests/postgres_tests/test_json.py [hunk 5]
```diff
@@ -19,6 +19,16 @@
...
 
 
+class TestModelMetaOrdering(PostgreSQLTestCase):
+
+    def test_ordering_by_json_field_value(self):
+        class TestJSONModel(JSONModel):
+            class Meta:
+                ordering = ['field__value', ]
+
+        self.assertFalse(TestJSONModel.check())
```
Comment (carltongibson) [inline tests/postgres_tests/test_json.py]: To match other model `check()` tests, use `self.assertEqual(TestJSONModel.check(), [])`

**Comment by hramezani** [comment]
@carltongibson requested changes are done

File: django/db/models/base.py [hunk 6]
```diff
@@ -1643,9 +1643,39 @@ def _check_ordering(cls):
...
+            if LOOKUP_SEP in f:
+                related_fields.append(f)
+            else:
+                _fields.append(f)
+        fields = iter(_fields)
+        related_fields = iter(related_fields)
+
+        # check related fields
+        for field in related_fields:
+            parts = field.split(LOOKUP_SEP)
```
Comment (felixxm) [inline django/db/models/base.py]: I would remove this temporary variable and move `split()` directly into loop
```python
for part in field.split(LOOKUP_SEP):
```

File: django/db/models/base.py [hunk 7]
```diff
@@ -1643,9 +1643,39 @@ def _check_ordering(cls):
...
+                except (FieldDoesNotExist, AttributeError):
+                    if fld is None or fld.get_transform(part) is None:
+                        errors.append(
+                            checks.Error(
+                                "'ordering' refers to the nonexistent related field or lookup '%s'." % field,
+                                obj=cls,
+                                id='models.E029',
+                            )
+                        )
+                continue
```
Comment (felixxm) [inline django/db/models/base.py]: I think `continue` is unnecessary.

File: django/db/models/base.py [hunk 8]
```diff
@@ -1643,9 +1643,39 @@ def _check_ordering(cls):
...
+            else:
+                _fields.append(f)
+        fields = iter(_fields)
+        related_fields = iter(related_fields)
+
+        # check related fields
+        for field in related_fields:
+            parts = field.split(LOOKUP_SEP)
+            _cls = cls
+            fld = None
```
Comment (felixxm) [inline django/db/models/base.py]: This line can be removed :thinking:.

File: tests/postgres_tests/test_json.py [hunk 9]
```diff
@@ -19,6 +19,16 @@
     pass
 
 
+class TestModelMetaOrdering(PostgreSQLTestCase):
+
+    def test_ordering_by_json_field_value(self):
+        class TestJSONModel(JSONModel):
+            class Meta:
+                ordering = ['field__value', ]
```
Comment (felixxm) [inline tests/postgres_tests/test_json.py]: Please remove trailing comma and space.

File: tests/invalid_models_tests/test_models.py [hunk 10]
```diff
@@ -665,6 +666,90 @@ class Meta:
...
+            Error(
+                "'ordering' refers to the nonexistent related field or lookup "
+                "'parent1__parent2__missing_field'.",
+                obj=Child,
+                id='models.E029',
+            )
+        ])
+
+    def test_ordering_allows_registered_lookups(self):
+
```
Comment (felixxm) [inline tests/invalid_models_tests/test_models.py]: Chop blank line.

File: django/db/models/base.py (hunk 8 above)
Comment (hramezani) [reply django/db/models/base.py]: This line cannot be removed, because if we will have an error on `fld = _cls._meta.get_field(part)` at the first round of loop. the `fld` variable on `if fld is None or fld.get_transform(part) is None:` line in `except` is `Unbounded` and we will get the `UnboundLocalError`

Comment (felixxm) [reply django/db/models/base.py]: You're right, thanks :+1:

File: django/db/models/base.py [hunk 11]
```diff
@@ -1643,9 +1643,37 @@ def _check_ordering(cls):
...
+        # Separate related field and non related fields.
+        _fields = []
+        related_fields = []
+        for f in fields:
+            if LOOKUP_SEP in f:
+                related_fields.append(f)
+            else:
+                _fields.append(f)
+        fields = iter(_fields)
+        related_fields = iter(related_fields)
```
Comment (felixxm) [inline django/db/models/base.py]: IMO this line is unnecessary, and above `iter()` call can be removed.

File: tests/postgres_tests/test_json.py [hunk 12]
```diff
@@ -19,6 +19,16 @@
     pass
 
 
+class TestModelMetaOrdering(PostgreSQLTestCase):
+
```
Comment (felixxm) [inline tests/postgres_tests/test_json.py]: Chop blank line.

**Comment by hramezani** [comment]
@felixxm Thanks for the comments.
requested changes are done

File: django/db/models/base.py [hunk 13]
```diff
@@ -1643,9 +1643,36 @@ def _check_ordering(cls):
...
+        for field in related_fields:
+            _cls = cls
+            fld = None
+            for part in field.split(LOOKUP_SEP):
+                try:
+                    fld = _cls._meta.get_field(part)
+                    if fld.is_relation:
+                        _cls = fld.get_path_info()[-1].to_opts.model
+                    else:
+                        _cls = None
```
Comment (felixxm) [inline django/db/models/base.py]: I wonder if we need:
```python
else:
    _cls = None
```
it looks unused at first glance and all tests pass without it :thinking: . Sorry for dripping comments.

**Comment by hramezani** [comment]
@felixxm Thank you again for the comment. Done!

**Comment by felixxm** [comment]
@hramezani Thanks for all updates :+1: I have one final comment :wink: I would reuse existing error `models.E015` instead of creating a new one. We can amend the message, e.g.
```
``ordering`` refers to the nonexistent field  ``<field name>``
```
to the
```
``ordering`` refers to the nonexistent field, related field or lookup  ``<field name>``
```

File: django/db/models/base.py [hunk 14]
```diff
@@ -1643,9 +1643,34 @@ def _check_ordering(cls):
...
+            for part in field.split(LOOKUP_SEP):
+                try:
+                    fld = _cls._meta.get_field(part)
+                    if fld.is_relation:
+                        _cls = fld.get_path_info()[-1].to_opts.model
+                except (FieldDoesNotExist, AttributeError):
+                    if fld is None or fld.get_transform(part) is None:
+                        errors.append(
+                            checks.Error(
+                                "'ordering' refers to the nonexistent field, related field or lookup '%s'." % field,
```
Comment (felixxm) [inline django/db/models/base.py]: Please wrap at 79 chars.

**Comment by hramezani** [comment]
@felixxm Done!

**Comment by hramezani** [comment]
@felixxm Done!

File: django/db/models/base.py [hunk 15]
```diff
@@ -1643,9 +1643,35 @@ def _check_ordering(cls):
         # Convert "-field" to "field".
         fields = ((f[1:] if f.startswith('-') else f) for f in fields)
 
-        # Skip ordering in the format field1__field2 (FIXME: checking
-        # this format would be nice, but it's a little fiddly).
-        fields = (f for f in fields if LOOKUP_SEP not in f)
+        # Separate related field and non related fields.
```
Comment (timgraham) [inline django/db/models/base.py]: related fields (add 's')
non-related (add dash)

File: django/db/models/base.py [hunk 16]
```diff
@@ -1643,9 +1643,35 @@ def _check_ordering(cls):
...
+                try:
+                    fld = _cls._meta.get_field(part)
+                    if fld.is_relation:
+                        _cls = fld.get_path_info()[-1].to_opts.model
+                except (FieldDoesNotExist, AttributeError):
+                    if fld is None or fld.get_transform(part) is None:
+                        errors.append(
+                            checks.Error(
+                                "'ordering' refers to the nonexistent field, "
+                                "related field or lookup '%s'." % field,
```
Comment (timgraham) [inline django/db/models/base.py]: Comma before 'or' per https://en.wikipedia.org/wiki/Serial_comma.

File: tests/postgres_tests/test_json.py [hunk 17]
```diff
@@ -19,6 +19,15 @@
     pass
 
 
+class TestModelMetaOrdering(PostgreSQLTestCase):
```
Comment (timgraham) [inline tests/postgres_tests/test_json.py]: This could be PostgreSQLSimpleTestCase I believe.

**Comment by felixxm** [comment]
@timgraham Thanks :bowing_man: I created follow up PR #11041.
